File: examples/trossen_stationary_ai/trossen_stationary_ai.cpp
/**
 * @file trossen_stationary_ai.cpp
 * @brief Bimanual AI Kit demo - 4 arms + RealSense cameras with config-driven setup
 *
 * Records all 4 arms (both leaders and followers) along with RealSense cameras.
 * All hardware parameters, session settings, and teleop setup are driven by a
 * single JSON config file with optional CLI overrides.
 *
 * Usage:
 *   ./trossen_stationary_ai [OPTIONS]
 *
 *   --config PATH       Path to robot config JSON
 *                       [default: examples/trossen_stationary_ai/config.json]
 *   --set KEY=VALUE     Override a config value using dot notation (repeatable)
 *   --dump-config       Print merged config as JSON and exit
 *   --help              Show this help and exit
 *
 * Examples:
 *   ./trossen_stationary_ai
 *   ./trossen_stationary_ai --config examples/trossen_stationary_ai/config.json
 *   ./trossen_stationary_ai --set hardware.arms.leader_left.ip_address=192.168.1.3
 *   ./trossen_stationary_ai --set session.max_duration=30 --set session.backend_type=lerobot
 *   ./trossen_stationary_ai --dump-config
 */

#include <chrono>
#include <filesystem>
#include <iostream>
#include <memory>
#include <string>
#include <thread>
#include <unordered_map>
#include <vector>

#include "libtrossen_arm/trossen_arm.hpp"
#include "nlohmann/json.hpp"
#include "trossen_sdk/configuration/cli_parser.hpp"
#include "trossen_sdk/configuration/loaders/json_loader.hpp"
#include "trossen_sdk/configuration/sdk_config.hpp"
#include "trossen_sdk/hw/arm/trossen_arm_producer.hpp"
#include "trossen_sdk/hw/arm/trossen_arm_component.hpp"
#include "trossen_sdk/hw/hardware_registry.hpp"
#include "trossen_sdk/io/backend_utils.hpp"
#include "trossen_sdk/runtime/producer_registry.hpp"
#include "trossen_sdk/runtime/push_producer_registry.hpp"
#include "trossen_sdk/runtime/session_manager.hpp"

#include "trossen_sdk/utils/app_utils.hpp"

static const std::vector<double> STAGED_POSITIONS = {
  0.0, 1.04719755, 0.523598776, 0.628318531, 0.0, 0.0, 0.0
};

static void print_usage(const char* program) {
  std::cout <<
    "Usage: " << program << " [OPTIONS]\n"
    "\n"
    "Options:\n"
    "  --config PATH      Path to robot config JSON\n"
    "                     [default: examples/trossen_stationary_ai/config.json]\n"
    "  --set KEY=VALUE    Override a config value using dot notation (repeatable)\n"
    "  --dump-config      Print merged config as JSON and exit\n"
    "  --help             Show this help and exit\n"
    "\n"
    "Examples:\n"
    "  " << program << "\n"
    "  " << program << " --config examples/trossen_stationary_ai/config.json\n"
    "  " << program << " --set hardware.arms.leader_left.ip_address=192.168.1.3\n"
    "  " << program << " --set session.max_duration=30\n"
    "  " << program << " --dump-config\n";
}

int main(int argc, char** argv) {
  trossen::configuration::CliParser cli(argc, argv);

  if (cli.has_flag("help")) {
    print_usage(argv[0]);
    return 0;
  }

  const std::string config_path =
    cli.get_string("config", "examples/trossen_stationary_ai/config.json");

  if (!std::filesystem::exists(config_path)) {
    std::cerr << "Error: config file not found: " << config_path << "\n";
    return 1;
  }

  // Load JSON and apply CLI overrides
  auto j = trossen::configuration::JsonLoader::load(config_path);
  const auto overrides = cli.get_set_overrides();
  if (!overrides.empty()) {
    j = trossen::configuration::merge_overrides(j, overrides);
  }

  if (cli.has_flag("dump-config")) {
    trossen::configuration::dump_config(j, "Trossen Stationary AI Kit Config");
    return 0;
  }

  // Parse unified robot config
  auto cfg = trossen::configuration::SdkConfig::from_json(j);

  // Populate GlobalConfig so SessionManager picks up session + backend settings
  cfg.populate_global_config();

  // Derive data root path from the TrossenMCAP backend config
  const std::string root = cfg.mcap_backend.root;

  // Derive per-type rates from the producer list
  float joint_rate_hz = 30.0f;
  float camera_fps = 30.0f;
  for (const auto& p : cfg.producers) {
    if (p.type == "trossen_arm") {
      joint_rate_hz = p.poll_rate_hz;
      break;
    }
  }
  for (const auto& p : cfg.producers) {
    if (p.type == "realsense_camera" || p.type == "opencv_camera" ||
        p.type == "zed_camera") {
      camera_fps = p.poll_rate_hz;
      break;
    }
  }

  std::vector<std::string> config_lines = {
    "Config file:          " + config_path,
    "Root directory:       " + root,
    "Backend:              " + cfg.session.backend_type,
    "Robot name:           " + cfg.robot_name
  };
  for (const auto& [id, arm] : cfg.hardware.arms) {
    config_lines.push_back(
      "Arm [" + id + "]:  " + arm.ip_address + " (" + arm.end_effector + ")");
  }
  config_lines.push_back("Joint rate:           " + std::to_string(joint_rate_hz) + " Hz");
  for (const auto& cam : cfg.hardware.cameras) {
    config_lines.push_back(
      "Camera [" + cam.id + "]:  " + cam.serial_number + "  " +
      std::to_string(cam.width) + "x" + std::to_string(cam.height) +
      " @ " + std::to_string(cam.fps) + " fps");
  }
  config_lines.push_back(
    "Teleop:               " +
    std::string(cfg.teleop.enabled ? "enabled" : "disabled") +
    " (" + std::to_string(cfg.teleop.pairs.size()) + " pairs)");

  trossen::utils::print_config_banner("Trossen Stationary AI Kit Demo Usage", config_lines);

  trossen::utils::install_signal_handler();
  std::filesystem::create_directories(root);

  // ──────────────────────────────────────────────────────────
  // Initialize arm hardware from config
  // ──────────────────────────────────────────────────────────

  std::cout << "Initializing hardware...\n";

  std::unordered_map<std::string,
    std::shared_ptr<trossen::hw::arm::TrossenArmComponent>> arm_components;

  for (const auto& [id, arm_cfg] : cfg.hardware.arms) {
    auto component = trossen::hw::HardwareRegistry::create(
      "trossen_arm", id, arm_cfg.to_json(), true);
    arm_components[id] =
      std::dynamic_pointer_cast<trossen::hw::arm::TrossenArmComponent>(component);
    std::cout << "  [ok] Arm [" << id << "] configured (" << arm_cfg.ip_address << ")\n";
  }

  // Stage all arms to starting positions
  const float moving_time_s = 2.0f;
  for (const auto& [id, comp] : arm_components) {
    auto driver = comp->get_hardware();
    driver->set_all_modes(trossen_arm::Mode::position);
    driver->set_all_positions(STAGED_POSITIONS, moving_time_s, false);
  }
  std::this_thread::sleep_for(std::chrono::duration<float>(moving_time_s + 0.1f));
  std::cout << "  [ok] All arms staged to starting positions\n";

  // Adjust gripper tolerance for follower arms
  for (const auto& pair : cfg.teleop.pairs) {
    auto it = arm_components.find(pair.follower);
    if (it == arm_components.end()) {
      continue;
    }
    auto driver = it->second->get_hardware();
    auto limits = driver->get_joint_limits();
    limits[driver->get_num_joints() - 1].position_tolerance = 0.01;
    driver->set_joint_limits(limits);
  }

  // ──────────────────────────────────────────────────────────
  // Session Manager + producers
  // ──────────────────────────────────────────────────────────

  trossen::runtime::SessionManager mgr;
  std::cout << "\nInitialized Session Manager\n";
  std::cout << "  Starting episode index: " << mgr.stats().current_episode_index << "\n";
  if (mgr.stats().current_episode_index > 0) {
    std::cout << "  (Resuming from existing episodes in directory)\n";
  }
  std::cout << "\n";

  // Pre-initialize camera hardware (keyed by camera id for producer lookup)
  std::unordered_map<std::string,
    std::shared_ptr<trossen::hw::HardwareComponent>> camera_components;
  std::unordered_map<std::string,
    const trossen::configuration::CameraConfig*> camera_cfg_map;
  for (const auto& cam_cfg : cfg.hardware.cameras) {
    auto cam_component = trossen::hw::HardwareRegistry::create(
      cam_cfg.type, cam_cfg.id, cam_cfg.to_json());
    camera_components[cam_cfg.id] = cam_component;
    camera_cfg_map[cam_cfg.id] = &cam_cfg;
    std::cout << "  [ok] Camera [" << cam_cfg.id << "] initialized ("
              << cam_cfg.serial_number << ")\n";
  }

  std::cout << "Creating producers...\n";

  // One producer per entry in the producers list
  for (const auto& prod_cfg : cfg.producers) {
    const auto period =
      std::chrono::milliseconds(static_cast<int>(1000.0f / prod_cfg.poll_rate_hz));

    if (prod_cfg.type == "trossen_arm") {
      auto prod = trossen::runtime::ProducerRegistry::create(
        "trossen_arm",
        arm_components.at(prod_cfg.hardware_id),
        prod_cfg.to_registry_json());
      mgr.add_producer(prod, period);
      std::cout << "  [ok] Arm producer [" << prod_cfg.stream_id << "] ("
                << prod_cfg.poll_rate_hz << " Hz)\n";
    } else if (camera_components.count(prod_cfg.hardware_id)) {
      const auto* cam = camera_cfg_map.at(prod_cfg.hardware_id);
      if (trossen::runtime::PushProducerRegistry::is_registered(prod_cfg.type)) {
        auto prod = trossen::runtime::PushProducerRegistry::create(
          prod_cfg.type,
          camera_components.at(prod_cfg.hardware_id),
          prod_cfg.to_registry_json(cam->width, cam->height, cam->fps));
        mgr.add_push_producer(prod);
        std::cout << "  [ok] Camera producer (push) [" << prod_cfg.stream_id << "] ("
                  << cam->width << "x" << cam->height << ")\n";
      } else {
        auto prod = trossen::runtime::ProducerRegistry::create(
          prod_cfg.type,
          camera_components.at(prod_cfg.hardware_id),
          prod_cfg.to_registry_json(cam->width, cam->height, cam->fps));
        mgr.add_producer(prod, period);
        std::cout << "  [ok] Camera producer [" << prod_cfg.stream_id << "] ("
                  << prod_cfg.poll_rate_hz << " Hz, "
                  << cam->width << "x" << cam->height << ")\n";
      }
    }
  }

  std::cout << "\nProducers registered. Ready to record.\n";

  // ──────────────────────────────────────────────────────────
  // Register lifecycle callbacks
  // ──────────────────────────────────────────────────────────

  // Count depth-capable cameras once for sanity checks
  int depth_cameras = 0;
  for (const auto& cam : cfg.hardware.cameras) {
    if (cam.use_depth) ++depth_cameras;
  }

  mgr.on_episode_ended([&](const trossen::runtime::SessionManager::Stats& stats) {
    const std::string file_path =
      trossen::utils::generate_episode_path(root, stats.current_episode_index);
    trossen::utils::print_episode_summary(file_path, stats);

    trossen::utils::SanityCheckConfig sanity_cfg{
      stats.elapsed.count(),
      static_cast<int>(cfg.hardware.arms.size()),
      joint_rate_hz,
      static_cast<int>(cfg.hardware.cameras.size()),
      static_cast<int>(camera_fps),
      5.0,
      depth_cameras
    };
    perform_sanity_check(stats.current_episode_index, stats.records_written_current, sanity_cfg);
  });

  mgr.on_pre_shutdown([&]() {
    std::cout << "\nReturning arms to starting positions...\n";
    for (const auto& [id, comp] : arm_components) {
      auto driver = comp->get_hardware();
      driver->set_all_modes(trossen_arm::Mode::position);
      driver->set_all_positions(STAGED_POSITIONS, moving_time_s, false);
    }
    std::this_thread::sleep_for(std::chrono::duration<float>(moving_time_s + 0.1f));

    std::cout << "Moving arms to sleep positions...\n";
    for (const auto& [id, comp] : arm_components) {
      auto driver = comp->get_hardware();
      driver->set_all_positions(
        std::vector<double>(driver->get_num_joints(), 0.0), moving_time_s, false);
    }
    std::this_thread::sleep_for(std::chrono::duration<float>(moving_time_s + 0.1f));

    std::cout << "Cleaning up arm drivers...\n";
    for (const auto& [id, comp] : arm_components) {
      comp->get_hardware()->cleanup();
    }
    std::cout << "  [ok] All arm drivers cleaned up\n";
  });

  // ──────────────────────────────────────────────────────────
  // Episode loop
  // ──────────────────────────────────────────────────────────

  while (true) {
    if (trossen::utils::g_stop_requested) {
      std::cout << "\n\nStopping at user request (Ctrl+C).\n";
      break;
    }

    if (cfg.teleop.enabled) {
      std::cout << "\nLocking leader arms and moving followers to match...\n";
      for (const auto& pair : cfg.teleop.pairs) {
        auto leader_it = arm_components.find(pair.leader);
        auto follower_it = arm_components.find(pair.follower);
        if (leader_it == arm_components.end() || follower_it == arm_components.end()) {
          continue;
        }
        auto leader_driver = leader_it->second->get_hardware();
        auto follower_driver = follower_it->second->get_hardware();
        leader_driver->set_all_modes(trossen_arm::Mode::position);
        follower_driver->set_all_modes(trossen_arm::Mode::position);
        follower_driver->set_all_positions(
          leader_driver->get_all_positions(), moving_time_s, false);
      }
      std::this_thread::sleep_for(std::chrono::duration<float>(moving_time_s + 0.1f));
      std::cout << "  [ok] Followers moved to match leaders\n";

      std::cout << "\n!! Enabling teleop mode !!\n";
      for (const auto& pair : cfg.teleop.pairs) {
        auto it = arm_components.find(pair.leader);
        if (it == arm_components.end()) {
          continue;
        }
        auto driver = it->second->get_hardware();
        driver->set_all_modes(trossen_arm::Mode::external_effort);
        driver->set_all_external_efforts(
          std::vector<double>(driver->get_num_joints(), 0.0), 0.0, false);
        std::cout << "   " << pair.leader << ": gravity compensation -> "
                  << pair.follower << " will mirror\n";
      }
      std::cout << "\n";
    }

    mgr.print_episode_header();

    if (!mgr.start_episode()) {
      break;
    }

    std::cout << "Recording...\n";

    // Teleop thread mirrors leader positions to followers at configured rate
    std::thread teleop_thread([&]() {
      if (!cfg.teleop.enabled) {
        return;
      }
      const auto sleep_ns = static_cast<int64_t>(1e9 / cfg.teleop.rate_hz);
      while (mgr.is_episode_active() && !trossen::utils::g_stop_requested) {
        for (const auto& pair : cfg.teleop.pairs) {
          auto leader_it = arm_components.find(pair.leader);
          auto follower_it = arm_components.find(pair.follower);
          if (leader_it == arm_components.end() || follower_it == arm_components.end()) {
            continue;
          }
          follower_it->second->get_hardware()->set_all_positions(
            leader_it->second->get_hardware()->get_all_positions(), 0.0f, false);
        }
        std::this_thread::sleep_for(std::chrono::nanoseconds(sleep_ns));
      }
    });

    mgr.monitor_episode();

    if (mgr.is_episode_active()) {
      mgr.stop_episode();
    }
    if (teleop_thread.joinable()) {
      teleop_thread.join();
    }

    if (trossen::utils::g_stop_requested) {
      std::cout << "\nStopping at user request (Ctrl+C).\n";
      break;
    }
  }

  // shutdown() calls stop_episode() (fires on_episode_ended) then on_pre_shutdown
  mgr.shutdown();

  const auto final_stats = mgr.stats();
  std::vector<std::string> extra_info = {
    "Data streams:         " +
      std::to_string(cfg.hardware.arms.size()) + " arms + " +
      std::to_string(cfg.hardware.cameras.size()) + " cameras"
  };
  trossen::utils::print_final_summary(
    final_stats.total_episodes_completed, root, extra_info);

  return 0;
}

File: python/trossen_sdk/demo/trossen_stationary_ai.py
# Copyright 2025 Trossen Robotics
#
# Purpose:
# Bimanual AI Kit data collection — 2 leader + 2 follower arms + 4 cameras.
# Python equivalent of examples/trossen_stationary_ai/trossen_stationary_ai.cpp.
#
# Hardware setup:
# 1. Four WXAI V0 arms (2 leaders + 2 followers) connected via Ethernet
# 2. Four RealSense cameras
#
# Usage:
#   python -m trossen_sdk.demo.trossen_stationary_ai
#   python -m trossen_sdk.demo.trossen_stationary_ai --config path/to/config.json

import argparse
import os

import trossen_sdk as ts


def main():
    parser = argparse.ArgumentParser(
        description="Trossen Stationary AI Data Collection")
    parser.add_argument(
        "--config",
        default="examples/trossen_stationary_ai/config.json",
        help="Path to robot config JSON",
    )
    args = parser.parse_args()

    if not os.path.exists(args.config):
        print(f"Error: config file not found: {args.config}")
        return 1

    # ── Load configuration ────────────────────────────────────────────────
    j = ts.JsonLoader.load(args.config)
    cfg = ts.SdkConfig.from_json(j)
    cfg.populate_global_config()

    root = cfg.mcap_backend.root

    config_lines = [
        f"Config file:          {args.config}",
        f"Root directory:       {root}",
        f"Backend:              {cfg.session.backend_type}",
        f"Robot name:           {cfg.robot_name}",
    ]
    for arm_id, arm_cfg in cfg.hardware.arms.items():
        config_lines.append(
            f"Arm [{arm_id}]:  {arm_cfg.ip_address} ({arm_cfg.end_effector})")
    for cam in cfg.hardware.cameras:
        config_lines.append(
            f"Camera [{cam.id}]:  {cam.serial_number}  "
            f"{cam.width}x{cam.height} @ {cam.fps} fps")
    config_lines.append(
        f"Teleop:               {'enabled' if cfg.teleop.enabled else 'disabled'} "
        f"({len(cfg.teleop.pairs)} pairs)")
    ts.print_config_banner("Trossen Stationary AI Kit (Python)", config_lines)

    ts.install_signal_handler()
    os.makedirs(root, exist_ok=True)

    # ── Initialize hardware ───────────────────────────────────────────────
    print("Initializing hardware...")

    arm_components = {}
    for arm_id, arm_cfg in cfg.hardware.arms.items():
        comp = ts.HardwareRegistry.create(
            "trossen_arm", arm_id, arm_cfg.to_json(), True)
        arm_components[arm_id] = comp
        print(f"  [ok] Arm [{arm_id}] ({arm_cfg.ip_address})")

    camera_components = {}
    camera_cfg_map = {}
    for cam_cfg in cfg.hardware.cameras:
        cam = ts.HardwareRegistry.create(
            cam_cfg.type, cam_cfg.id, cam_cfg.to_json())
        camera_components[cam_cfg.id] = cam
        camera_cfg_map[cam_cfg.id] = cam_cfg
        print(f"  [ok] Camera [{cam_cfg.id}] ({cam_cfg.serial_number})")

    # ── Teleop controllers ────────────────────────────────────────────────
    # Resolves teleop pairs from the global teleop config + ActiveHardwareRegistry.
    # Hardware components above were created with mark_active=True (the default),
    # so they're already discoverable by ID.
    controllers = ts.create_teleop_controllers_from_global_config()

    # ── Session Manager + producers ───────────────────────────────────────
    mgr = ts.SessionManager()
    print(f"\nSession Manager (starting episode: "
          f"{mgr.stats().current_episode_index})")

    for prod_cfg in cfg.producers:
        period_ms = int(1000.0 / prod_cfg.poll_rate_hz)

        if prod_cfg.type == "trossen_arm":
            prod = ts.ProducerRegistry.create(
                "trossen_arm", arm_components[prod_cfg.hardware_id],
                prod_cfg.to_registry_json())
            mgr.add_producer(prod, period_ms)
            print(f"  [ok] Arm producer [{prod_cfg.stream_id}] "
                  f"({prod_cfg.poll_rate_hz} Hz)")

        elif prod_cfg.hardware_id in camera_components:
            cam = camera_cfg_map[prod_cfg.hardware_id]
            rj = prod_cfg.to_registry_json_camera(cam.width, cam.height, cam.fps)
            if ts.PushProducerRegistry.is_registered(prod_cfg.type):
                prod = ts.PushProducerRegistry.create(
                    prod_cfg.type, camera_components[prod_cfg.hardware_id], rj)
                mgr.add_push_producer(prod)
                print(f"  [ok] Camera (push) [{prod_cfg.stream_id}] "
                      f"({cam.width}x{cam.height})")
            else:
                prod = ts.ProducerRegistry.create(
                    prod_cfg.type, camera_components[prod_cfg.hardware_id], rj)
                mgr.add_producer(prod, period_ms)
                print(f"  [ok] Camera [{prod_cfg.stream_id}] "
                      f"({prod_cfg.poll_rate_hz} Hz)")

    # ── Lifecycle callbacks ───────────────────────────────────────────────
    mgr.on_pre_episode(lambda: (_start_controllers(controllers), True)[-1])
    mgr.on_episode_started(
        lambda: print("Episode started - recording active."))
    mgr.on_episode_ended(lambda stats: (
        _stop_controllers(controllers),
        ts.print_episode_summary(
            ts.generate_episode_path(root, stats.current_episode_index), stats),
    ))
    mgr.on_pre_shutdown(lambda: _stop_controllers(controllers))

    # ── Episode loop ──────────────────────────────────────────────────────
    print("\nReady to record.\n")

    while not ts.get_stop_requested():
        mgr.print_episode_header()

        if not mgr.start_episode():
            break

        action = mgr.monitor_episode(print_stats=True)

        if action == ts.UserAction.kReRecord:
            mgr.discard_current_episode()
            continue

        if mgr.is_episode_active():
            mgr.stop_episode()

        if ts.get_stop_requested():
            break

        action = mgr.wait_for_reset()
        if action == ts.UserAction.kStop:
            break
        if action == ts.UserAction.kReRecord:
            mgr.discard_last_episode()

    mgr.shutdown()
    stats = mgr.stats()
    ts.print_final_summary(
        stats.total_episodes_completed, root,
        [f"Data streams: {len(cfg.hardware.arms)} arms + "
         f"{len(cfg.hardware.cameras)} cameras"])


def _start_controllers(controllers):
    for ctrl in controllers:
        ctrl.prepare_teleop(); ctrl.teleop()


def _stop_controllers(controllers):
    for ctrl in controllers:
        if ctrl.is_running():
            ctrl.stop_teleop()


if __name__ == "__main__":
    main()
